src/swap/comit.rs

```diff
@@ -0,0 +1,346 @@
+    EC: LatestBlock<Block = ethereum::Block>
+        + BlockByHash<Block = ethereum::Block, BlockHash = ethereum::Hash>
+        + btsieve::ethereum::ReceiptByHash,
+{
+    let hbit_funded = match alice.execute_fund(&hbit_params).await {
+        Ok(hbit_funded) => hbit_funded,
+        Err(_) => return Ok(()),
+    };
+
+    let herc20_deployed =
+        match herc20::watch_for_deployed(ethereum_connector, herc20_params.clone(), start_of_swap)
+            .await
+        {
+            Ok(herc20_deployed) => herc20_deployed,
+            Err(_) => {
+                alice.execute_refund(hbit_params, hbit_funded).await?;
+
+                return Ok(());
+            }
+        };
+
+    let _herc20_funded = match herc20::watch_for_funded(
+        ethereum_connector,
+        herc20_params.clone(),
+        start_of_swap,
+        herc20_deployed.clone(),
+    )
+    .await
+    {
+        Ok(herc20_funded) => herc20_funded,
+        Err(_) => {
+            alice.execute_refund(hbit_params, hbit_funded).await?;
+
+            return Ok(());
+        }
+    };
+
+    let _herc20_redeemed = match alice
+        .execute_redeem(herc20_params, secret, herc20_deployed, start_of_swap)
+        .await
+    {
+        Ok(herc20_redeemed) => herc20_redeemed,
+        Err(_) => {
+            alice.execute_refund(hbit_params, hbit_funded).await?;
+
+            return Ok(());
+        }
+    };
+
+    let hbit_redeem = hbit::watch_for_redeemed(
+        bitcoin_connector,
+        &hbit_params.shared,
+        hbit_funded.location,
+        start_of_swap,
+    );
+    let hbit_refund = alice.execute_refund(hbit_params, hbit_funded);
+
+    pin_mut!(hbit_redeem);
+    pin_mut!(hbit_refund);
+
+    match future::select(hbit_redeem, hbit_refund).await {
+        Either::Left((Ok(_hbit_redeemed), _)) => Ok(()),
+        Either::Right((Ok(_hbit_refunded), _)) => Ok(()),
+        Either::Left((Err(_), hbit_refund)) => {
+            hbit_refund.await?;
+            Ok(())
+        }
+        Either::Right((Err(_), _hbit_redeem)) => Ok(()),
+    }
+}
+
+/// Execute Bob's side of a Hbit<->Herc20 swap.
+pub async fn hbit_herc20_bob<B, BC, EC>(
+    bob: B,
+    bitcoin_connector: &BC,
+    ethereum_connector: &EC,
+    hbit_params: hbit::Params,
+    herc20_params: herc20::Params,
+    start_of_swap: NaiveDateTime,
+) -> anyhow::Result<()>
+where
+    B: herc20::ExecuteDeploy + herc20::ExecuteFund + herc20::ExecuteRefund + hbit::ExecuteRedeem,
+    BC: LatestBlock<Block = ::bitcoin::Block>
+        + BlockByHash<Block = ::bitcoin::Block, BlockHash = ::bitcoin::BlockHash>,
+    EC: LatestBlock<Block = ethereum::Block>
+        + BlockByHash<Block = ethereum::Block, BlockHash = ethereum::Hash>
+        + btsieve::ethereum::ReceiptByHash,
+{
+    let hbit_funded =
+        match hbit::watch_for_funded(bitcoin_connector, &hbit_params.shared, start_of_swap).await {
+            Ok(hbit_funded) => hbit_funded,
+            Err(_) => return Ok(()),
+        };
```

Comment:
This is exactly the code pattern that `?` expands to except that `?` returns an error. What do you think of this:

```suggestion
    let hbit_funded = hbit::watch_for_funded(bitcoin_connector, &hbit_params.shared, start_of_swap).await.context("hbit not funded")?;
```

Reply:
That's a great idea!